## BST 
In this notebook I'll implement an insert and breadth-first traversal functions for a  binary search tree

In [1]:
# Node class with value and 2 children
class Node:
    def __init__(self, val):
        self.left = None
        self.right = None
        self.val = val  

In [2]:
# Insert new nodes
def insert(root, node):
    # if it's the first node, it's the root
    if root is None:
        root = node

    else:
        # if the value is smaller or equal to the root; go left
        if node.val <= root.val:
            if root.left is None:
                root.left = node
            else:
                insert(root.left, node)
        # the value is larger than the root's value
        else:
            if root.right is None:
                root.right = node
            else:
                insert(root.right, node)

In [3]:
# breadth-first traversal using queue
def print_tree(root):
    # if tree is empty
    if root is None:
        return
    # create queue and add the root
    queue = []
    queue.append(root)
    
    # will go over the all queue and print each value in FIFO order
    while (len(queue) > 0):
        # pop out the first element and print it
        node = queue.pop(0)
        print(node.val)
        
        # if have children, add them to the queue
        if node.left is not None:
            queue.append(node.left)
            
        if node.right is not None:
            queue.append(node.right)
        

In [4]:
# to generate "random" numbers
import random
random.seed(a=3)

# Create random value list.
rand_list = list(random.sample(range(10000), 15))

# list to store the nodes
node_list = []

# creates 15 new node variables (called node0, node2,..., node14) with value from the random list
for i in range(15):
    exec("node%s = Node(rand_list[%s])" % (i, i))

# create and insert to the tree
insert(None,node0)
insert(node0, node1)
insert(node0, node2)
insert(node0, node3)
insert(node0, node4)
insert(node0, node5)
insert(node0, node6)
insert(node0, node7)
insert(node0, node8)
insert(node0, node9)
insert(node0, node10)
insert(node0, node11)
insert(node0, node12)
insert(node0, node13)
insert(node0, node14)

In [5]:
print_tree(node0)

3898
2136
9709
1073
3839
8916
9894
215
6061
9516
9922
4249
7766
9024
7687


## What the output should be like
for this random list: <br>
3898,9709,8916,2136,6061,9894,7766,9516,1073,9922,215,7687,4249,9024,3839 <br>
The output should be like (left to right): <br>
Level 0: 3898 (root) <br>
Level 1: 2136, 9709 <br>
Level 2: 1037, 3839, 8916, 9894 <br>
Level 3: 215, 6061, 9516, 9922 <br>
Level 4: 4249, 7766, 9024 <br>
Level 5: 7687 <br>